Mathlib/LinearAlgebra/Matrix/Irreducible/Defs.lean

```diff
@@ -0,0 +1,253 @@
+      have h_Am : 0 < (A ^ m) i l := by
+        by_contra h
+        have h_eq : (A ^ m) i l = 0 := le_antisymm (le_of_not_gt h) hAm_nonneg
+        simp [h_eq] at hl_pos
+      have h_A : 0 < A l j := by
+        by_contra h
+        have h_eq : A l j = 0 := le_antisymm (le_of_not_gt h) hA_nonneg'
+        simp [h_eq] at hl_pos
+      obtain ⟨⟨p, hp_len⟩⟩ := (ih i l).mp h_Am
+      exact ⟨letI := toQuiver A; p.cons h_A, by
+          subst hp_len
+          simp only [Path.length_cons]⟩
+    · rintro ⟨p, hp_len⟩
+      cases p with
+      | nil => simp [Quiver.Path.length] at hp_len
+      | @cons b _ q e =>
+        simp only [Quiver.Path.length_cons, Nat.succ.injEq] at hp_len
+        have h_Am_pos : 0 < (A ^ m) i b := (ih i b).mpr ⟨q, hp_len⟩
+        let h_A_pos := e
+        have h_prod : 0 < (A ^ m) i b * A b j := mul_pos h_Am_pos h_A_pos
+        exact
+          (Finset.sum_pos_iff_of_nonneg
+            (fun x _ => mul_nonneg (pow_apply_nonneg hA m i x) (hA x j))).2
+            ⟨b, Finset.mem_univ b, h_prod⟩
+
+/-- Irreducibility of a nonnegative matrix `A` is equivalent to entrywise positivity of some
+power: between any two indices `i, j` there exists a positive integer `k` such that the
+`(i, j)`-entry of `A ^ k` is strictly positive. -/
+theorem IsIrreducible.iff_exists_pow_pos
+    [IsOrderedRing R] [PosMulStrictMono R] [Nontrivial R] [DecidableEq n]
+    (hA : ∀ i j, 0 ≤ A i j) :
+    IsIrreducible A ↔ ∀ i j, ∃ k > 0, 0 < (A ^ k) i j := by
+  constructor
+  · intro h_irr i j
+    letI : Quiver n := toQuiver A
+    obtain ⟨p, hp_len⟩ := h_irr.2 i j
+    refine ⟨p.length, hp_len, ?_⟩
+    have : Nonempty {q : Path i j // q.length = p.length} := ⟨⟨p, rfl⟩⟩
+    have hpos :=
+      (pow_entry_pos_iff_exists_path (A := A) hA p.length i j).2 this
+    simpa using hpos
+  · intro h_exists
+    constructor
+    · exact hA
+    · intro i j
+      obtain ⟨k, hk_pos, hk_entry⟩ := h_exists i j
+      letI : Quiver n := toQuiver A
+      obtain ⟨⟨p, hp_len⟩⟩ :=
+        (pow_entry_pos_iff_exists_path (A := A) hA k i j).mp hk_entry
+      subst hp_len
+      exact ⟨p, hk_pos⟩
+
+/-- If a nonnegative square matrix `A` is primitive, then `A` is irreducible. -/
+theorem IsPrimitive.to_IsIrreducible
+    [IsOrderedRing R] [PosMulStrictMono R] [Nontrivial R] [DecidableEq n]
+    (h_prim : IsPrimitive A) : IsIrreducible A := by
+  obtain ⟨h_nonneg, k, hk_pos, hk_all⟩ := h_prim
+  refine ⟨h_nonneg, fun i j ↦ ?_⟩
+  letI : Quiver n := toQuiver A
+  have ⟨⟨p, hp⟩⟩ : Nonempty {p : Path i j // p.length = k} :=
+    (pow_entry_pos_iff_exists_path h_nonneg k i j).mp (hk_all i j)
+  exact ⟨p, hp ▸ hk_pos⟩
+
+/-! ## Transposition -/
+
+/-- Reverse a path in `toQuiver A` to a path in `toQuiver Aᵀ`, swapping endpoints. -/
+def transposeRev
+    {n R : Type*} [Ring R] [LinearOrder R] {A : Matrix n n R} :
+    ∀ {i j : n}, (@Quiver.Path n (toQuiver A) i j) →
+      (@Quiver.Path n (toQuiver Aᵀ) j i) := by
+  letI : Quiver n := toQuiver A
+  intro i j p
```

Comment:
Name and [left of colon](https://leanprover-community.github.io/contribute/style.html#hypotheses-left-of-colon) style tweak:
```suggestion
def transposePath {i j : n} (p : @Quiver.Path n A.toQuiver i j) :
    @Quiver.Path n Aᵀ.toQuiver j i := by
```